File: backend/app/schemas/text.py
"""
Text schemas for API request/response validation.
"""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.models.text import DocumentType, TextFormat
from app.models.text_word_link import TextWordLinkStatus


class TextBase(BaseModel):
    """Base schema for Text"""

    title: str = Field(..., min_length=1, max_length=500)
    content: str = Field(..., min_length=1)
    document_type: DocumentType
    format: TextFormat = TextFormat.PLAIN
    language_id: UUID | None = None  # Language of the content
    document_id: UUID | None = None  # Parent document
    is_primary: bool = False
    source: str | None = Field(None, max_length=500)
    notes: str | None = None
    # Editor pin for the guided learning path (null = computed placement).
    learning_order: int | None = None


class TextCreate(TextBase):
    """Schema for creating a new text"""

    pass


class TextUpdate(BaseModel):
    """Schema for updating a text"""

    title: str | None = Field(None, min_length=1, max_length=500)
    content: str | None = Field(None, min_length=1)
    document_type: DocumentType | None = None
    format: TextFormat | None = None
    language_id: UUID | None = None
    document_id: UUID | None = None
    is_primary: bool | None = None
    source: str | None = Field(None, max_length=500)
    notes: str | None = None
    learning_order: int | None = None


class TextInDB(TextBase):
    """Schema for text as stored in database"""

    id: UUID
    created_by_id: UUID
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class Text(TextInDB):
    """Complete text schema"""

    pass


class TextWordLinkBase(BaseModel):
    """Base schema for a word link within a text"""

    start_char: int = Field(..., ge=0, description="Start character offset (inclusive)")
    end_char: int = Field(..., gt=0, description="End character offset (exclusive)")
    notes: str | None = Field(None, max_length=500)


class TextWordLinkCreate(TextWordLinkBase):
    """Schema for creating a link to an existing word"""

    word_form_id: UUID
    status: TextWordLinkStatus = TextWordLinkStatus.CONFIRMED


class TextWordLinkUpdate(BaseModel):
    """Schema for updating link metadata or status"""

    status: TextWordLinkStatus | None = None
    notes: str | None = Field(None, max_length=500)
    word_form_id: UUID | None = None


class TextWordLink(BaseModel):
    """Complete schema for a link between text span and word"""

    id: UUID
    text_id: UUID
    word_form_id: UUID
    start_char: int
    end_char: int
    status: TextWordLinkStatus
    confidence: float | None = None
    notes: str | None = None
    created_by_id: UUID | None = None
    verified_by_id: UUID | None = None
    created_at: datetime
    updated_at: datetime
    verified_at: datetime | None = None
    # Form-level info (the actual surface form the link points at)
    word_text: str | None = None
    word_form_romanization: str | None = None
    word_form_ipa: str | None = None
    word_form_notes: str | None = None
    # Lexeme-level info (the dictionary entry the form belongs to)
    lexeme_id: UUID | None = None
    word_lemma: str | None = None
    word_part_of_speech: str | None = None
    word_notes: str | None = None
    word_language_id: UUID | None = None

    model_config = ConfigDict(from_attributes=True)


class TextWithLinks(Text):
    """Text schema including link metadata"""

    word_links: list[TextWordLink] = []


class TextListItem(BaseModel):
    """Simplified schema for text lists"""

    id: UUID
    title: str
    content: str = Field(..., max_length=200)  # Truncated content for lists
    document_type: DocumentType
    language_id: UUID | None = None
    document_id: UUID | None = None
    is_primary: bool
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class TextFilter(BaseModel):
    """Schema for filtering and searching texts"""

    document_type: DocumentType | None = None
    language_id: UUID | None = None
    document_id: UUID | None = None
    is_primary: bool | None = None
    search_term: str | None = None  # Search in content or title
    created_by_id: UUID | None = None
    skip: int = Field(0, ge=0)
    limit: int = Field(100, ge=1, le=1000)


class TextStatistics(BaseModel):
    """Statistics about texts in the database"""

    total_texts: int
    texts_by_type: dict[str, int]
    texts_by_language: dict[str, int]
    texts_with_linked_words: int
    primary_texts: int

File: frontend/src/types/text.ts
/**
 * Text type definitions
 *
 * Texts contain actual content in specific languages.
 * Multiple Texts can belong to the same Document (translations).
 */

export enum DocumentType {
  // Full documents
  STORY = 'story',
  HISTORICAL_RECORD = 'historical_record',
  BOOK = 'book',
  ARTICLE = 'article',
  TRANSCRIPTION = 'transcription',
  WRITING_STANDARD = 'writing_standard',

  // Text snippets
  DEFINITION = 'definition',
  LITERAL_TRANSLATION = 'literal_translation',
  CONTEXT_NOTE = 'context_note',
  USAGE_EXAMPLE = 'usage_example',
  ETYMOLOGY = 'etymology',
  CULTURAL_SIGNIFICANCE = 'cultural_significance',
  TRANSLATION = 'translation',
  NOTE = 'note',

  OTHER = 'other',
}

export enum TextWordLinkStatus {
  SUGGESTED = 'suggested',
  CONFIRMED = 'confirmed',
  REJECTED = 'rejected',
}

/** How a Text's content should be rendered. Plain prose vs. markdown. */
export enum TextFormat {
  PLAIN = 'plain',
  MARKDOWN = 'markdown',
}

/** Publication state: suggestions from non-editors start pending_review. */
export enum TextStatus {
  PENDING_REVIEW = 'pending_review',
  PUBLISHED = 'published',
  ARCHIVED = 'archived',
}

export interface TextWordLink {
  id: string;
  text_id: string;
  word_form_id: string;
  start_char: number;
  end_char: number;
  status: TextWordLinkStatus;
  confidence?: number | null;
  notes?: string | null;
  created_by_id?: string | null;
  verified_by_id?: string | null;
  created_at: string;
  updated_at: string;
  verified_at?: string | null;
  // Surface form (matching word_form_id)
  word_text?: string | null;
  word_form_romanization?: string | null;
  word_form_ipa?: string | null;
  word_form_notes?: string | null;
  // Parent lexeme (the dictionary entry)
  lexeme_id?: string | null;
  word_lemma?: string | null;
  word_part_of_speech?: string | null;
  word_notes?: string | null;
  word_language_id?: string | null;
}

export interface TextWordLinkCreate {
  word_form_id: string;
  start_char: number;
  end_char: number;
  status?: TextWordLinkStatus;
  notes?: string;
}

export interface TextWordLinkUpdate {
  status?: TextWordLinkStatus;
  notes?: string;
  word_form_id?: string;
}

export interface Text {
  id: string;
  title: string;
  content: string;
  document_type: DocumentType;
  format?: TextFormat;
  status?: TextStatus;
  language_id?: string;
  document_id?: string;
  is_primary: boolean;
  source?: string;
  notes?: string;
  /** Editor pin for the guided learning path (null = computed placement). */
  learning_order?: number | null;
  created_by_id: string;
  created_at: string;
  updated_at: string;
  word_links?: TextWordLink[];
}

export interface TextListItem {
  id: string;
  title: string;
  content: string; // May be truncated
  document_type: DocumentType;
  language_id?: string;
  document_id?: string;
  is_primary: boolean;
  created_at: string;
}

export interface TextWithLinks extends Text {
  word_links: TextWordLink[];
  /** Confirmed-link coverage (0..1); populated by the document endpoints. */
  link_coverage?: number | null;
}

export interface TextCreate {
  title: string;
  content: string;
  document_type: DocumentType;
  language_id?: string;
  document_id?: string;
  is_primary?: boolean;
  source?: string;
  notes?: string;
}

export interface TextUpdate {
  title?: string;
  content?: string;
  document_type?: DocumentType;
  format?: TextFormat;
  language_id?: string;
  document_id?: string;
  is_primary?: boolean;
  source?: string;
  notes?: string;
  learning_order?: number | null;
}

export interface TextFilter {
  document_type?: DocumentType;
  language_id?: string;
  document_id?: string;
  is_primary?: boolean;
  search_term?: string;
  created_by_id?: string;
  skip?: number;
  limit?: number;
}

export interface TextStatistics {
  total_texts: number;
  texts_by_type: Record<string, number>;
  texts_by_language: Record<string, number>;
  texts_with_linked_words: number;
  primary_texts: number;
}
